runners/direct-java/src/main/java/org/apache/beam/runners/direct/EvaluationContext.java

```diff
@@ -183,6 +188,15 @@ public void initialize(
         committedResult.getUnprocessedInputs().orElse(null),
         committedResult.getOutputs(),
         result.getWatermarkHold());
+
+    // Callback and requested bundle finalizations
+    for (InMemoryBundleFinalizer.Finalization finalization : result.getBundleFinalizations()) {
+      try {
+        finalization.getCallback().onBundleSuccess();
+      } catch (Exception e) {
+        LOG.warn("Failed to finalize requested bundle {}", finalization, e);
```

Comment:
I don't think logging only the finalization tells us enough. Can we log more information about the bundle itself?

Reply:
Added the completed bundle to the output.